### PR TITLE
Start enforcing typecheck

### PR DESCRIPTION
**Pull request by EarlMilktea**
**Description of the change:**

Use `mypy` and `pyright` to detect type warnings.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -93,3 +93,31 @@ convention = "numpy"
 [tool.pytest.ini_options]
 # Silence cotengra warning
 filterwarnings = ["ignore:Couldn't import `kahypar`"]
+
+[tool.mypy]
+disable_error_code = [
+  "type-arg",
```
Comment (thierry-martinez) [inline pyproject.toml:99]: We can enable `type-arg` with only few changes: https://github.com/thierry-martinez/graphix/commit/d5c4d50d653082294b0eb93784cb8da46ec7145b

File: pyproject.toml [hunk 2]
```diff
@@ -93,3 +93,31 @@ convention = "numpy"
...
+  "unused-ignore", # Necessary for pyright
```
Comment (thierry-martinez) [inline pyproject.toml:100]: Can you explain why this line is necessary? I don't see any error from either `mypy` or `pyright` when it is removed.

**Review by thierry-martinez — CHANGES_REQUESTED** [review]
Very nice! However, I think we can remove the disabled error code.

File: requirements.txt [hunk 3]
```diff
@@ -9,4 +9,6 @@ pydantic
 quimb>=1.4.0
 scipy
 sympy>=1.9
+types-networkx
+types-psutil
```
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply requirements.txt:13]: Sounds reasonable, but especially for `types-networkx` affecting our public APIs, I'm not sure whether users can see annotations without installing it...

File: pyproject.toml (hunk 1 above)
Comment (EarlMilktea) [reply pyproject.toml:99]: I know it's not necessary for now, but would be necessary as it's really hard to resolve all the placeholders used in `npt.NDArray`: this is why I added it.
How do you think about it?

File: requirements.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (EarlMilktea) [reply requirements.txt]: Looks like it's unnecessary.
Fixed in 2bfc76e21f97cb133193d986abc549684e0a87e3.

File: pyproject.toml (hunk 2 above)
Comment (EarlMilktea) [reply pyproject.toml:100]: This can be removed if we use `pyright: ignore` instead of `type: ignore` when only `pyright` complains.

Comment (EarlMilktea) [reply pyproject.toml]: Fixed in 74ce7925275807dd18f21b301f3f659b6d5dc0f1.

File: pyproject.toml (hunk 1 above)
Comment (EarlMilktea) [reply pyproject.toml]: Anyway let me adopt your idea here.

**Review by thierry-martinez — APPROVED** [review]

**Comment by thierry-martinez** [comment]
LGTM! Thanks.

**Review by shinich1 — APPROVED** [review]
Looks good! let's squash and merge.

**Pull request merged**